docs/twemcache-rs.md

```diff
@@ -0,0 +1,126 @@
+the session is established, it is passed over to the worker thread which then
+handles network IO, request parsing, and response composition.
+
+We found the Rust implementation was competitive with the C implementation in
+terms of throughput and latency, giving us the confidence to proceed with
+converting the Twemcache backend to Rust.
+
+## Twemcache in Rust
+
+The standard Pelikan Twemcache backend uses slab based storage. But with the
+development of the Pelikan Segcache backend, we found that for many workloads
+there was a significant benefit in being able to proactively remove expired
+items. Based on the benefits seen with the development of Segcache, we decided
+to focus our efforts on implementing the Rust Twemcache-compatible backend using
+a Segcache storage design.
+
+### Segcache Storage
+
+The Rust implementation of Segcache attempts to closely follow the C
+implementation. We have three main components, the hashtable, segments
+containing items, and TTL buckets to enable proactive expiration. However,
+unlike the C implementation, we use a "SegCache" struct to contain all of the
+components and retain ownership over the data. This allows us to define our
+high-level interface and keep the implementation details separate.
+
+### Threading
+
+In testing, we found that the Rust implementation performed well compared to
+Pelikan Twemcache and single-threaded Twemcache. However, we have some cache
+instances that have multi-threaded Twemcache configurations. While the Rust
+implementation typically outperformed the single-threaded Twemcache + TLS, it
+could not keep up handling the TLS streams compared to multi-threaded Twemcache.
+
+As an experiment, we tried an alternative threading architecture to avoid making
+the storage layer concurrent. Since the bottleneck is in the network IO and TLS
+handling, we can divide the established sessions across multiple worker threads.
+These worker threads can then communicate with a shared storage thread which
+will own the datastructure and handle requests.
+
+```
+┌─────────────┐                         ┌─────────────┐                                             
+│             │                         │             │                                             
+│    Admin    │                         │             │    ┌────────────────┐                       
+│             │                    ┌───▶│   Worker    │◀──▶│SPSC Queue Pair │◀──┐                   
+│    :9999    │                    │    │             │    └────────────────┘   │   ┌──────────────┐
+│             │                    │    │             │                         │   │              │
+└─────────────┘                    │    └─────────────┘                         │   │              │
+                                   │    ┌─────────────┐                         ├──▶│   Storage    │
+┌─────────────┐                    │    │             │                         │   │              │
```

Comment:
I can redraw it with something like this for the middle:
```
                           ┌────────────┐
                   ╭───────▶  worker 1  │
                   │       └────────────┘
                   │       ┌────────────┐
┌──────────────┐   │╭──────▶  worker 2  │
│              │───╯│      └────────────┘
│              │────╯      ┌────────────┐
│   sessions   │───────────▶  worker 3  │
│              │           └────────────┘
│              │───╮                     
└──────────────┘   │   ┋         ┋       
                   │                     
                   │       ┌────────────┐
                   ╰───────▶  worker n  │
                           └────────────┘
```

Reply:
Yeah - this doc needs work. If you have feels about the diagram, feel free to drop something pretty here =)